### PR TITLE
#167727632 Build out endpoint for user profile creation and update

### PR DESCRIPTION
**Pull request by chrismeeky**
### What does this PR do?
- Lets user create or update his profile

#### Description of Task to be completed?

- Create an endpoint where user can create/update his profile once he completes registration. Users input should be validated using sequelize's built-in validator. User should be re-routed to the request page upon successful registration. Refer to the screenshot below for fields required for the profile creation.

- Create an endpoint such that when a user is redirected to the profile settings page, a fetch API call is made to this endpoint to pull user's current settings. Changes are applied through the _profile update endpoint_  and the user redirected to the request page upon updating his profile.

- Create a _middleware_ that ensures that the user gets his profile created once he registers on the app and  also ensures that if a user with an incomplete profile returns to login, he is reminded/redirected to complete his profile

- Protect routes using JWT authentication

#### How should this be manually tested?

To test the two endpoints for _fetching current profile settings_ and _updating profile settings_ ,
* Clone the repo and `cd` into the project directory
* Run `yarn install` to install the project's dependencies
* Create the environment variables in a _.env_ file and set the environment to _development_
* Run `yarn migrate` to create tables
* Run `yarn dev-start` to start the app
* Make a GET request to localhost:3000/api/v1/profile/<:UserId>  to get the user's profile
Screenshot from 2019-08-22 14-20-20
* Make a PATCH request to localhost:3000/api/v1/profile/<:UserId>  to update user's profile with fields as shown in the picture below.
Screenshot from 2019-08-22 14-21-36
#### What are the relevant pivotal tracker stories?
[#167727632](https://www.pivotaltracker.com/n/projects/2381976)

### CODE REVIEW

**Review by minlinx — COMMENTED** [review]
Great job!!! I dont think it is necessary to **import {sequelize, Sequelize }** in the controller since we have done so in the **models/index**

File: src/index.js [hunk 1]
```diff
@@ -2,13 +2,10 @@ import express from 'express';
 import bodyParser from 'body-parser';
 import cors from 'cors';
 import swaggerUI from 'swagger-ui-express';
-import dotenv from 'dotenv';
 import morgan from 'morgan';
 import doc from '../doc.json';
 import routes from './routes/v1';
 
-dotenv.config();
```
Comment (ChidozieGabriel) [inline src/index.js]: Please, why did you remove line 10? dotenv.config() needs to be called at the entry file, once and for all.
Check out this [conversation](https://github.com/andela/rogue-backend/pull/23#discussion_r316191703) . Thanks.

File: src/utils/helper.js [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    const date = new Date(newdate);
+    return date.toLocaleDateString();
+  }
+  return oldDate;
+};
+
+export const confirmUser = (req, res) => {
+  const { decodedPayload, success } = verifyToken(getToken(req));
+  if (!success) {
+    return res.status(401).json({
```
Comment (ChidozieGabriel) [inline src/utils/helper.js]: Authorization has already been handled. To secure a route, use the `checkToken` middleware  in `src/middlewares/authorization.js` file. Here is a sample:
```
...
import { checkToken } from '../../middlewares/authorization';
...

router.patch('/user/:id', checkToken, Profile.updateProfile);
router.get('/user/:id', checkToken, Profile.getProfile);
```

> It works like magic :-)

Hence, you may not  need to create a `getToken` function.

File: src/routes/v1/index.js [hunk 3]
```diff
@@ -1,9 +1,12 @@
 import express from 'express';
 import authRoute from './authRoute';
+import profileRoute from './ProfileRoute';
 import requestRoute from './requestRoute';
 
 const router = express.Router();
 router.use('/auth', authRoute);
+router.use('/auth', profileRoute);
+router.use('/request', requestRoute);
 router.use('/request', requestRoute);
```
Comment (ChidozieGabriel) [inline src/routes/v1/index.js]: Line 9 is repeated in line 10. Please do delete one of them.

File: src/routes/v1/ProfileRoute.js [hunk 4]
```diff
@@ -0,0 +1,13 @@
+import express from 'express';
+import Profile from '../../controllers/ProfileUpdateController';
+import profileChecker from '../../middlewares/CheckProfile';
+
+const router = express.Router();
+
+router.post('/signup_test', Profile.create);
+router.post('/signin_test', profileChecker, Profile.signInUser);
+router.patch('/user/:id', Profile.updateProfile);
+router.get('/user/:id', Profile.getProfile);
```
Comment (ChidozieGabriel) [inline src/routes/v1/ProfileRoute.js]:
```suggestion
router.get('/user/:id', checkToken, Profile.getProfile);
```

File: src/routes/v1/ProfileRoute.js [hunk 5]
```diff
@@ -0,0 +1,13 @@
+import express from 'express';
+import Profile from '../../controllers/ProfileUpdateController';
+import profileChecker from '../../middlewares/CheckProfile';
+
+const router = express.Router();
+
+router.post('/signup_test', Profile.create);
+router.post('/signin_test', profileChecker, Profile.signInUser);
+router.patch('/user/:id', Profile.updateProfile);
```
Comment (ChidozieGabriel) [inline src/routes/v1/ProfileRoute.js]:
```suggestion
router.patch('/user/:id', checkToken, Profile.updateProfile);
```

File: src/routes/v1/ProfileRoute.js [hunk 6]
```diff
@@ -0,0 +1,13 @@
+import express from 'express';
+import Profile from '../../controllers/ProfileUpdateController';
+import profileChecker from '../../middlewares/CheckProfile';
+
```
Comment (ChidozieGabriel) [inline src/routes/v1/ProfileRoute.js]:
```suggestion
import { checkToken } from '../../middlewares/authorization';

```

**Review by ChidozieGabriel — CHANGES_REQUESTED** [review]
@chrismeeky

> Your lines of code are super-awesome :100: 

Please do consider the following review. Thanks.

File: src/index.js (hunk 1 above)
Comment (chrismeeky) [reply src/index.js]: Thank you. I have corrected that.

**Review by okikiola11 — COMMENTED** [review]
This is looking good, There are no bugs.

**Review by okikiola11 — DISMISSED** [review]

File: src/routes/v1/ProfileRoute.js (hunk 5 above)
Comment (chrismeeky) [reply src/routes/v1/ProfileRoute.js]: Thank you. All feedback has been implemented

File: src/index.js [hunk 7]
```diff
@@ -2,8 +2,8 @@ import express from 'express';
 import bodyParser from 'body-parser';
 import cors from 'cors';
 import swaggerUI from 'swagger-ui-express';
-import dotenv from 'dotenv';
 import morgan from 'morgan';
+import dotenv from 'dotenv';
```
Comment (ChidozieGabriel) [inline src/index.js]: Did you call dotenv.config() here? If _'no'_ please do.

**Review by ChidozieGabriel — CHANGES_REQUESTED** [review]
I have seen your implementation. Great :+1: 
However, consider this review. Thanks.

File: src/routes/v1/ProfileRoute.js [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+import profileChecker from '../../middlewares/CheckProfile';
+import checkToken from '../../middlewares/authorization';
+import confirmUser from '../../middlewares/ConfirmUser';
+
+const router = express.Router();
+
+router.post('/signup_test', Profile.create);
+router.post('/signin_test', profileChecker, Profile.signInUser);
+router.patch('/user/:id', checkToken, confirmUser, Profile.updateProfile);
+router.get('/user/:id', checkToken, confirmUser, Profile.getProfile);
```
Comment (ChidozieGabriel) [inline src/routes/v1/ProfileRoute.js]: What does `confirmUser` do that `checkToken` does not? Could you explain please?

**Review by ChidozieGabriel — CHANGES_REQUESTED** [review]

File: src/controllers/ProfileUpdateController.js [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+  static async getProfile(req, res) {
+    const userId = parseInt(req.params.id, 10);
+    try {
+      const user = await Users.findOne({ where: { id: userId } });
+      ResponseHandler.success(res, user);
+    } catch (error) {
+      ResponseHandler.error(res, 'No user with that id was found', 404);
+    }
+  }
+}
```
Comment (Daymorelah) [inline src/controllers/ProfileUpdateController.js]: I see no need for creating a `ProfileUpdateController.js`. if at all you want to implement your controller in a separate file, you name the file `userController.js`. Also, check the way I import models into controllers. Also, check the way I import files from the utility folder. you don't just create your own style, you follow the style the codebase is using.

File: src/middlewares/authorization.js [hunk 10]
```diff
@@ -1,20 +1,19 @@
...
-  const token = req.headers['x-access-token']
-    || req.query.token
-    || req.headers.authorization;
-  if (token) {
+const checkToken = (req, res, next) => {
+  const queryToken = req.query.token;
+  const accessToken = req.headers['x-access-token'];
+  const bearerToken = req.headers.authorization;
+  if (queryToken || accessToken || bearerToken) {
+    const token = queryToken || accessToken || bearerToken.split(' ')[1];
```
Comment (Daymorelah) [inline src/middlewares/authorization.js]: was it necessary to edit did lines 4-9?. was the original code that was there causing a bug in the app?. please, revert what you dod here.

File: src/models/User.js [hunk 11]
```diff
@@ -40,16 +35,113 @@ export default (sequelize, DataTypes) => {
...
+          args: [6],
+          msg: 'Line manager must be at least 6 characters long.'
+        },
+      }
+    },
+    isVerified: DataTypes.BOOLEAN,
+    slug: {
+      type: DataTypes.STRING,
+      allowNull: true,
+    },
```
Comment (Daymorelah) [inline src/models/User.js:179]: Add the `allowNull:false` atrribute to all new columns you added.

File: src/routes/v1/requestRoute.js [hunk 12]
```diff
@@ -1,6 +1,6 @@
 import express from 'express';
 import { bookTrip } from '../../controllers/RequestController';
-import { checkToken } from '../../middlewares/authorization';
+import checkToken from '../../middlewares/authorization';
```
Comment (Daymorelah) [inline src/routes/v1/requestRoute.js]: Why did you edit this line?

File: src/utils/ResponseHandler.js [hunk 13]
```diff
@@ -18,7 +18,7 @@ export default class ResponseHandler {
    * @returns {void}
    */
   static success(res, data, status = 200) {
-    res.status(status).json({
+    return res.status(status).json({
```
Comment (Daymorelah) [inline src/utils/ResponseHandler.js:21]: I believe i have implemented this method before. why the repetition?

File: src/utils/ResponseHandler.js [hunk 14]
```diff
@@ -36,7 +36,7 @@ export default class ResponseHandler {
    * @returns {void}
    */
   static error(res, message = 'Bad request', status = 400) {
-    res.status(status).json({
+    return res.status(status).json({
```
Comment (Daymorelah) [inline src/utils/ResponseHandler.js:39]: Ditto!

File: src/utils/authentication.js [hunk 15]
```diff
@@ -11,6 +11,11 @@ export const generateToken = (payload, expiresIn = '24h') => {
 };
 
 export const verifyToken = token => {
+  if (!token) {
+    return {
+      success: false
+    };
+  }
```
Comment (Daymorelah) [inline src/utils/authentication.js]: are the lines 14-18 necessary?

File: src/utils/helper.js [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+  const domainCom = domains[randomNumber4];
+
+  const firstString = alphabets[randomNumber1];
+  const secondString = alphabets[randomNumber2];
+  const thirdString = alphabets[randomNumber3];
+  // eslint-disable-next-line max-len
+  const userEmail = `${firstString}${secondString}${thirdString}${randomNumber1}${randomNumber2}@${domainCom}`;
+  return userEmail;
+};
+
```
Comment (Daymorelah) [inline src/utils/helper.js]: is this function necessary?

File: src/routes/v1/ProfileRoute.js (hunk 8 above)
Comment (chrismeeky) [reply src/routes/v1/ProfileRoute.js]: confirm user checks if the ID from the decoded token is the same with the one passed in the `req.params`

**Review by Daymorelah — CHANGES_REQUESTED** [review]
Please implement this feature. All your feature requires is to create a simple controller, that handle a put/patch request to update a role in the user's table. @victor-shagor already created a file `users.js` in the controller folder u can use it.

File: src/controllers/ProfileUpdateController.js (hunk 9 above)
Comment (chrismeeky) [reply src/controllers/ProfileUpdateController.js]: I will remove the extra controllers. I didn't want them to be a blocker as I needed the extra endpoints to test the feature and planned to revert once the extra endpoints have been merged. Thanks for setting the standards ! I will modify.

File: src/middlewares/authorization.js (hunk 10 above)
Comment (chrismeeky) [reply src/middlewares/authorization.js]: I thought we might need an extra option for passing the token

File: src/models/User.js (hunk 11 above)
Comment (chrismeeky) [reply src/models/User.js:179]: All right

File: src/utils/ResponseHandler.js (hunk 13 above)
Comment (chrismeeky) [reply src/utils/ResponseHandler.js:21]: I didn't re-implement. I only added a return statement to avoid setting the header twice

File: src/utils/authentication.js (hunk 15 above)
Comment (chrismeeky) [reply src/utils/authentication.js]: It is a check to make sure token exists before the `shuffleToken` method is called. The shuffleToken has a `split` method that crashes the app if what it's  splitting is undefined

File: src/utils/helper.js (hunk 16 above)
Comment (chrismeeky) [reply src/utils/helper.js]: Removed. Thanks

File: src/index.js (hunk 7 above)
Comment (chrismeeky) [reply src/index.js]: Yes I did

File: src/controllers/UserController.js [hunk 17]
```diff
@@ -147,6 +147,77 @@ class UserController {
...
+   * @static
+   * @param {object} req HTTP Request object
+   * @param {object} res HTTP Response object
+   * @returns {object} HTTP Response object
+   * @memberof ProfileController
+   */
+  static async updateProfile(req, res) {
+    const userId = req.params.id;
+    try {
+      const user = await User.findOne({ where: { id: userId } });
```
Comment (Daymorelah) [inline src/controllers/UserController.js]: Since we are looking for an id which is the primary key, use the `findByPk` method instead. check it out in the sequelize docs.

File: src/controllers/UserController.js [hunk 18]
```diff
@@ -147,6 +147,77 @@ class UserController {
...
+          city: req.body.city || user.city,
+          state: req.body.state || user.state,
+          zip: req.body.zip || user.zip,
+          country: req.body.country || user.country,
+          profileImage: req.body.profileImage || user.profileImage,
+          role: req.body.role || user.role,
+          department: req.body.department || user.department,
+          lineManager: req.body.lineManager || user.lineManager,
+          hasProfile: true,
+        });
```
Comment (Daymorelah) [inline src/controllers/UserController.js]: hmm, so lines 165-184 does not look efficient. find a way to only update what the user is sending to the backend. with your current implementation if we add a new column, then we need to come here to add that column when we want to update. please refactor. Reach out if you need help in coming up with the logic for it.

File: src/controllers/UserController.js [hunk 19]
```diff
@@ -147,6 +147,77 @@ class UserController {
...
+      } catch (error) {
+        return HelperMethods.clientError(res, error, 400);
```
Comment (Daymorelah) [inline src/controllers/UserController.js]: why are you sending the whole error to the `clientError` helper method?, I believe this method is expecting a client error message, not an error object. Please look at how the helper function works.

File: src/controllers/UserController.js [hunk 20]
```diff
@@ -147,6 +147,77 @@ class UserController {
...
+      }
+      return HelperMethods
+        .requestSuccessful(res, {
+          success: true,
+          userDetails: user,
```
Comment (Daymorelah) [inline src/controllers/UserController.js]: Don't send all the user details to the front end. also, add a message property that tells the use the request was successful. look at the sequelize docs, use the `returns` option to return only necessary columns from the backend after updating, that is if sending the user details is necessary.

File: src/controllers/UserController.js [hunk 21]
```diff
@@ -147,6 +147,77 @@ class UserController {
...
+        }, 200);
+    } catch (err) {
```
Comment (Daymorelah) [inline src/controllers/UserController.js]: why are you having more than one try-catch in a controller?. There should be only one try-catch per controller.

File: src/controllers/UserController.js [hunk 22]
```diff
@@ -147,6 +147,77 @@ class UserController {
...
+      return HelperMethods.clientError(res, 'User does not exist', 404);
```
Comment (Daymorelah) [inline src/controllers/UserController.js:172]: Also, i think it should be a server error that would cause the code to enter here, not a client error.

File: src/middlewares/authorization.js [hunk 23]
```diff
@@ -27,6 +27,26 @@ class Authorization {
...
+   * @param {object} res - Response object
+   * @param {callback} next - The callback that passes the request
+   * to the next handler
+   * @returns {callback} next - The callback that passes the request
+   * to the next handler
+   * @returns {object} res - Response object containing an error due
+   * to unauthorized user
+   */
+  static async confirmUser(req, res, next) {
+    if (req.decoded.id !== req.params.id) {
```
Comment (Daymorelah) [inline src/middlewares/authorization.js]: 👍

File: src/middlewares/authorization.js [hunk 24]
```diff
@@ -27,6 +27,26 @@ class Authorization {
...
+      return HelperMethods.clientError(res,
+        'you are not authorized to perform this action',
```
Comment (Daymorelah) [inline src/middlewares/authorization.js]: I feel `you cannot perform this action` is more descriptive

File: src/models/User.js [hunk 25]
```diff
@@ -57,6 +57,106 @@ export default (sequelize, DataTypes) => {
...
+      type: DataTypes.STRING,
+      validate: {
+        len: {
+          args: [4],
+          msg: 'Gender must be at least 4 characters long.'
+        },
+      }
+    },
+    birthdate: {
+      type: DataTypes.STRING,
```
Comment (Daymorelah) [inline src/models/User.js]: In the user-migration file, you set `birthdate` to have a datatype of DATE, but in here you set it as `STRING`, please use the appropriate one.

File: src/models/User.js (hunk 11 above)
Comment (Daymorelah) [reply src/models/User.js:179]: sorry, revert my feedback!

File: src/middlewares/authorization.js (hunk 23 above)
Comment (Daymorelah) [reply src/middlewares/authorization.js]: I love this.

File: src/controllers/UserController.js [hunk 26]
```diff
@@ -147,6 +147,77 @@ class UserController {
...
+   * @static
+   * @param {object} req client request
+   * @param {object} res server response
+   * @returns {object} server response object
+   * @memberof ProfileController
+   */
+  static async getProfile(req, res) {
+    const userId = req.params.id;
+    try {
+      const user = await User.findOne({ where: { id: userId } });
```
Comment (Daymorelah) [inline src/controllers/UserController.js]: Use `findByPk` instead of `findOne`.

File: src/controllers/UserController.js [hunk 27]
```diff
@@ -147,6 +147,77 @@ class UserController {
...
+      HelperMethods
```
Comment (Daymorelah) [inline src/controllers/UserController.js]: Check if the request returned a user before sending a success message. check the user controller for a hint on how t go about this.

File: src/controllers/UserController.js [hunk 28]
```diff
@@ -147,6 +147,77 @@ class UserController {
...
+        .requestSuccessful(res, {
+          success: true,
+          userDetails: user,
+        }, 200);
+    } catch (error) {
+      return HelperMethods.clientError(res, 'User does not exist', 404);
```
Comment (Daymorelah) [inline src/controllers/UserController.js]: We want to send a client error to the frontend when the user is not found. I feel here, we should send a server error.

File: src/test/integrationTests/userController.test.js [hunk 29]
```diff
@@ -143,4 +143,137 @@ describe('Integration tests for the user controller', () => {
...
+        preferredCurrency: 'USD',
+        profileImage: 'https://profileurl.com',
+        city: 'Enugu',
+        state: 'Enugu',
+        zip: '154587',
+        country: 'Nigeria',
+        role: 'SuperAdmin',
+        department: 'Accounting',
+        lineManager: 'Daniels',
+      };
```
Comment (Daymorelah) [inline src/test/integrationTests/userController.test.js:148]: implementing my feedback about making the update logic more effective will not prevent listing all these columns.

File: src/test/integrationTests/userController.test.js [hunk 30]
```diff
@@ -143,4 +143,137 @@ describe('Integration tests for the user controller', () => {
...
+          .equal(newData.zip);
+        expect(result.body.data.userDetails.country).to
+          .equal(newData.country);
+        expect(result.body.data.userDetails.role).to
+          .equal(newData.role);
+        expect(result.body.data.userDetails.department).to
+          .equal(newData.department);
+        expect(result.body.data.userDetails.lineManager).to
+          .equal(newData.lineManager);
+      });
```
Comment (Daymorelah) [inline src/test/integrationTests/userController.test.js:157]: It would also prevent this long list of assertions. you can only test to update two or three columns or even just one and it would suffice.

File: src/test/integrationTests/userController.test.js [hunk 31]
```diff
@@ -143,4 +143,137 @@ describe('Integration tests for the user controller', () => {
...
```
Comment (Daymorelah) [inline src/test/integrationTests/userController.test.js:198]: Remember not to send the user's password. There is a method that should make you select only specific fields/columns.

File: src/utils/ResponseHandler.js (hunk 13 above)
Comment (Daymorelah) [reply src/utils/ResponseHandler.js:21]: Please delete this file. All response helper methods are already declared in the helpermethod.js file

File: src/utils/helperMethod.js [hunk 32]
```diff
@@ -73,6 +73,20 @@ class HelperMethods {
...
+   * @param {object} oldDate - This is User's current date of birth
+   * @returns {string} date - Contains the date of birth
+   */
+  static returnDate(newDate, oldDate) {
+    if (newDate) {
+      const date = new Date(newDate);
+      return date.toLocaleDateString();
+    }
+    return oldDate;
+  }
```
Comment (Daymorelah) [inline src/utils/helperMethod.js]: I do not see any sequelize validation error being returned here as your method description says in its JSDOC. if you want to validate the date, then create a method in the validation file for validating the date being passed from the request.

**Review by Daymorelah — CHANGES_REQUESTED** [review]

**Review by Daymorelah — APPROVED** [review]
LGTM!
